Add support for embedding pretty printers via `#[debugger_visualizer]` attribute

ridwanabdillahi opened the pull request:
Initial support for [RFC 3191](https://github.com/rust-lang/rfcs/pull/3191) in PR https://github.com/rust-lang/rust/pull/91779 was scoped to supporting embedding NatVis files using a new attribute. This PR implements the pretty printer support as stated in the RFC mentioned above.

This change includes embedding pretty printers in the `.debug_gdb_scripts` just as the pretty printers for rustc are embedded today. Also added additional tests for embedded pretty printers. Additionally cleaned up error checking so all error checking is done up front regardless of the current target.

RFC: https://github.com/rust-lang/rfcs/pull/3191

rust-highfive commented:
r? @michaelwoerister

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6430859683?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/f8cfc2dd10f3eb637eeff5328b4a89527fbee564/checks/6430859683/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] src/test/ui/feature-gates/feature-gate-debugger-visualizer.rs stdout ----
diff of stderr:

7    = note: see issue #95939 <https://github.com/rust-lang/rust/issues/95939> for more information
8    = help: add `#![feature(debugger_visualizer)]` to the crate attributes to enable
- error: aborting due to previous error
- error: aborting due to previous error
+ error: invalid file specified for `#[debugger_visualizer]` attribute.
+    |
+    |
+ LL | #![debugger_visualizer(natvis_file = "../foo.natvis")]
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
+    |
+    = note: visualizer file specified does not exist or could not be read.
+ 
+ error: aborting due to 2 previous errors
---
To only update this specific test, also pass `--test-args feature-gates/feature-gate-debugger-visualizer.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/feature-gates/feature-gate-debugger-visualizer.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-debugger-visualizer" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gates/feature-gate-debugger-visualizer/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0658]: the `#[debugger_visualizer]` attribute is an experimental feature
   |
   |
LL | #![debugger_visualizer(natvis_file = "../foo.natvis")] //~ ERROR the `#[debugger_visualizer]` attribute is an experimental feature
   |
   = note: see issue #95939 <https://github.com/rust-lang/rust/issues/95939> for more information
   = note: see issue #95939 <https://github.com/rust-lang/rust/issues/95939> for more information
   = help: add `#![feature(debugger_visualizer)]` to the crate attributes to enable

error: invalid file specified for `#[debugger_visualizer]` attribute.
   |
   |
LL | #![debugger_visualizer(natvis_file = "../foo.natvis")] //~ ERROR the `#[debugger_visualizer]` attribute is an experimental feature
   |
   = note: visualizer file specified does not exist or could not be read.

error: aborting due to 2 previous errors
```

</details>

luqmana commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -891,14 +911,11 @@ impl CrateInfo {
-                }
-            }
+        if tcx.sess.target.is_like_msvc {
+            info.debugger_visualizers =
```
Worth renaming to `natvis_debugger_visualizers` to make it clear we only care for those in terms of what the linker is given later?

ridwanabdillahi replied:
This sounds good, I've renamed the field to express it's only accounting for NatVis debugger visualizers.

luqmana commented on src/doc/unstable-book/src/language-features/debugger-visualizer.md:
```diff
@@ -14,12 +14,8 @@ to embed a debugger visualizer file into the PDB/ELF generated by `rustc`.
-
-## Limitations
-
-Currently, this feature only supports embedding Natvis files on `-windows-msvc`
```
The `msvc` part is still true right? It doesn't seem like this would work with the `-windows-gnu` targets so worth keeping language along those lines.

ridwanabdillahi commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -2005,12 +2001,36 @@ impl CheckAttrVisitor<'_> {
+
+        if file.is_file() && std::fs::read(&file).is_ok() {
+            true
+        } else {
```
I removed the `is_file()` check since it's redundant but I'd like to read the file anyway. Thus, if the file cannot be read, this can be caught at an early stage as opposed to waiting for the query to be used which might not happen for certain crates even if the attribute is used.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6459983191?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/223d6ba35fac8b0caf7fb169a86f1074354d1d1d/checks/6459983191/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [ui] src/test/ui/invalid/invalid-debugger-visualizer-option.rs stdout ----
diff of stderr:

8    = note: OR
9    = note: expected: `gdb_script_file = "..."`
- error: The system cannot find the file specified. (os error 2)
+ error: No such file or directory (os error 2)
12   --> $DIR/invalid-debugger-visualizer-option.rs:3:24
13    |
13    |
14 LL | #![debugger_visualizer(natvis_file = "../foo.random")]
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-debugger-visualizer-option/invalid-debugger-visualizer-option.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-debugger-visualizer-option/invalid-debugger-visualizer-option.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args invalid/invalid-debugger-visualizer-option.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/invalid/invalid-debugger-visualizer-option.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-debugger-visualizer-option" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-debugger-visualizer-option/auxiliary"
stdout: none
--- stderr -------------------------------
  --> /checkout/src/test/ui/invalid/invalid-debugger-visualizer-option.rs:2:24
   |
   |
LL | #![debugger_visualizer(random_file = "../foo.random")] //~ ERROR invalid argument
   |
   |
   = note: expected: `natvis_file = "..."`
   = note: OR
   = note: expected: `gdb_script_file = "..."`
error: No such file or directory (os error 2)
  --> /checkout/src/test/ui/invalid/invalid-debugger-visualizer-option.rs:3:24
   |
   |
LL | #![debugger_visualizer(natvis_file = "../foo.random")] //~ ERROR

error: aborting due to 2 previous errors
------------------------------------------

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #96825) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

michaelwoerister reviewed (commented):
Thanks for the PR, @ridwanabdillahi! It looks very good already (it's great that it extends testing), but we'll need to resolve the issue around the linkage of `__rustc_debug_gdb_scripts_section__` somehow (see below).

michaelwoerister commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -1979,46 +1980,67 @@ impl CheckAttrVisitor<'_> {
+                }
+            };
+
+        match std::fs::read(&file) {
```
I just saw that you discussed this with @luqmana already. How about [File::open](https://doc.rust-lang.org/std/fs/struct.File.html#method.open)? That should also make sure that the file is readable.

ridwanabdillahi commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -486,6 +487,30 @@ fn get_argc_argv<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
+    tcx: TyCtxt<'_>,
+    visualizer_type: DebuggerVisualizerType,
+) -> FxHashSet<DebuggerVisualizerFile> {
+    if tcx.sess.crate_types().iter().all(|crate_type| crate_type.is_archive()) {
```
@michaelwoerister 

No, I don't believe we do since the visualizers will still live in the crate metadata for the static libs. In the case of MSVC targets, the visualizers can only be embedded in the PDB which does not get created for a static lib. I'm not as familiar with linux targets but does this work differently? Is it sufficient to only embed the transitive visualizers in a leaf crate such as an executable, since they will be linking against all of the static libs anyway.

ridwanabdillahi replied:
As per my previous comment, I'll reach out to offline to resolve this issue as well.

michaelwoerister replied:
The name `CrateType::StaticLib` might be a bit misleading. Static libs are not Rust libraries anymore and contain no crate metadata. One uses them if one needs to produce a static library that looks like a C library from the outside (similar to `cdylib`). See https://doc.rust-lang.org/reference/linkage.html.

So as far as the Rust compiler is concerned, static libs *are* leaf crates. In the case of PDBs, there's not much we can do, I think. The build system that will eventually invoke the linker with the static lib as input needs to take care of adding the relevant NatVis files (unless there's some feature that allows to bundle NatVis files in regular `.lib` files).

But for ELF binaries we can just add the pretty printers as a normal section of the object files.

ridwanabdillahi commented on compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs:
```diff
@@ -37,9 +40,27 @@ pub fn get_or_insert_gdb_debug_scripts_section_global<'ll>(cx: &CodegenCx<'ll, '
 
```
@michaelwoerister 

Hmmm, I see. Prior to this change all object files generated by rustc would have the same definition for the `__rustc_debug_gdb_scripts_section__` section which would be `b"\x01gdb_load_rust_pretty_printers.py\0"`. I believe my change would be problematic in that case. We should discuss how to best work around this issue. I'll reach out to you offline.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6566265672?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/3be81f76e876b77fa54e0a5a302bba815cff7b9a/checks/6566265672/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between ee160f2f5e73b6f5954bc33f059c316d9e8582c4 and 69a738cf3f6a86f20a047fc7f11ef35f6491eb2e
Submodules were updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
Verifying status of edition-guide...
Verifying status of rls...
This PR updated 'src/tools/rls', verifying if status is 'test-pass'...

We detected that this PR updated 'rls', but its tests failed.

If you do intend to update 'rls', please check the error messages above and
commit another update.

If you do NOT intend to update 'rls', please ensure you did not accidentally
change the submodule at 'src/tools/rls'. You may ask your reviewer for the
Build completed unsuccessfully in 0:00:00
```

</details>

michaelwoerister commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -882,14 +902,17 @@ impl CrateInfo {
+            && tcx.sess.crate_types().iter().any(|&crate_type| {
+                crate_type == CrateType::Executable
+                    || crate_type == CrateType::Dylib
+                    || crate_type == CrateType::Cdylib
```
For a condition like this it's useful to use an exhaustive match statement, so that we don't miss updating it if a new crate type gets added. It also makes it easy to add documentation for the various cases. E.g.:
```rust 
tcx.sess.crate_types().iter().any(|&crate_type| match {
  CrateType::Executable | CrateType::Dylib | crate_type == CrateType::Cdylib => {
    // These are crate types for which we invoke the linker
    true
  }
  CrateType::ProcMacro => {
    // We could embed NatVis for proc macro crates too (to improve the debugging experience for them)
    // but it does not seem like a good default, since this is a rare use case and we don't want to slow down
    // the common case.
    false
  }
  CrateType::Staticlib | CrateType::Rlib => {
    // We don't invoke the linker for these, so we don't need to collect the NatVis for them.
    false
  }
})
```

ridwanabdillahi committed:
Add support for embedding pretty printers via the `#[debugger_visualizer]` attribute. Add tests for embedding pretty printers and update documentation.

Ensure all error checking for `#[debugger_visualizer]` is done up front and not when the `debugger_visualizer` query is run.

Clean up potential ODR violations when embedding pretty printers into the `__rustc_debug_gdb_scripts_section__` section.

Respond to PR comments and update documentation.

ridwanabdillahi committed:
Update documentation.

michaelwoerister commented on compiler/rustc_passes/src/debugger_visualizer.rs:
```diff
@@ -34,45 +33,28 @@ fn check_for_debugger_visualizer<'tcx>(
+                    debugger_visualizers
+                        .insert(DebuggerVisualizerFile::new(Arc::from(contents), visualizer_type));
+                }
+                _ => {}
```
I don't think we should silently ignore the error case here. Let's just emit an error.

michaelwoerister reviewed (commented):
Thanks, @ridwanabdillahi, looks great!

Please address the error message issue mentioned below, then this should be ready to go.

ridwanabdillahi committed:
Respond to PR comments.

michaelwoerister commented:
Thanks, @ridwanabdillahi! This looks good to me now.

@bors r+

bors commented:
:pushpin: Commit 7ac62ce75cef963017245287f0a7b140e09589d7 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 7ac62ce75cef963017245287f0a7b140e09589d7 -->
<!-- homu: {"type":"Approved","sha":"7ac62ce75cef963017245287f0a7b140e09589d7","approver":"michaelwoerister"} -->

ridwanabdillahi commented:
@rustbot label +F-debugger_visualizer
